File: src/main/java/edu/osu/cse6341/lispInterpreter/singleton/ParserSingleton.java
package edu.osu.cse6341.lispInterpreter.singleton;

import edu.osu.cse6341.lispInterpreter.parser.Parser;
import lombok.Getter;

@Getter
public enum ParserSingleton {
    INSTANCE;

    private final Parser parser;

    ParserSingleton() {
        parser = Parser.newInstance(
            AsserterSingleton.INSTANCE.getTokenKindAsserter(),
            GeneratorSingleton.INSTANCE.getNodeGenerator()
        );
    }
}

File: src/main/java/com/soapdogg/lispInterpreter/singleton/ParserSingleton.kt
package com.soapdogg.lispInterpreter.singleton

import com.soapdogg.lispInterpreter.parser.*

enum class ParserSingleton {
    INSTANCE;

    val expressionListNodeParser: ExpressionListNodeParser
    val nodeParser: NodeParser
    val rootParser: RootParser

    init {
        expressionListNodeParser = ExpressionListNodeParser(
            GeneratorSingleton.INSTANCE.nodeGenerator
        )
        nodeParser = NodeParser(
            expressionListNodeParser
        )
        rootParser = RootParser(
            nodeParser
        )
    }
}